secureboot_policy.go

```diff
@@ -865,42 +915,58 @@ func (g *secureBootPolicyGen) computeAndExtendVerificationMeasurement(paths []*s
 			dbs = append(dbs, p.dbSet.mokDb, p.dbSet.shimDb)
 		}
 
-		var root *efiSignatureData
-		var rootDb *secureBootDb
-	Outer:
-		for _, db := range dbs {
-			if db == nil {
-				continue
-			}
+		var authority *secureBootAuthority
 
-			for _, s := range db.signatures {
-				// Ignore signatures that aren't X509 certificates
-				if s.signatureType != *efiCertX509Guid {
+		// To determine what CA certificate will be used to authenticate this image, iterate over the signatures in the order in which they
+		// appear in the binary in this outer loop. Iterating over the CA certificates occurs in an inner loop. This behaviour isn't defined
+		// in the UEFI specification but it matches EDK2 and the firmware on the Intel NUC. If an implementation iterates over the CA
+		// certificates in an outer loop and the signatures in an inner loop, then this may produce the wrong result.
+		for _, sig := range sigs {
+			for _, db := range dbs {
+				if db == nil {
 					continue
 				}
 
-				if bytes.Equal(s.data, signer.Raw) {
-					// The signing certificate is actually the root in the DB
-					root = s
-					rootDb = db
-					break Outer
-				}
+				for _, caSig := range db.signatures {
+					// Ignore signatures that aren't X509 certificates
+					if caSig.signatureType != *efiCertX509Guid {
+						continue
+					}
 
-				c, err := x509.ParseCertificate(s.data)
-				if err != nil {
-					continue
+					ca, err := x509.ParseCertificate(caSig.data)
+					if err != nil {
+						continue
+					}
+
+					roots := x509.NewCertPool()
+					roots.AddCert(ca)
+
+					opts := x509.VerifyOptions{
+						Intermediates: sig.intermediates,
+						Roots:         roots,
+						KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny}}
+					if _, err := sig.signer.Verify(opts); err == nil {
+						// This signer certificate is trusted by this authority
+						authority = &secureBootAuthority{signature: caSig, source: db}
+						break
+					}
+
+					if authority != nil {
+						break
+					}
```

Comment:
do we need this check at this level?  also wouldn't  using  break Label as the original code be simpler?  I am probably missing something

Reply:
I've added the label back now